docs(overview): implement newly organzied overview page

TheNando opened the pull request:
https://jira.rax.io/browse/FRMW-27
https://jira.rax.io/browse/FRMW-24
https://jira.rax.io/browse/FRMW-47

TheNando commented:
*Overview Page*
![image](https://cloud.githubusercontent.com/assets/6663710/8419152/40cd0b3c-1e7d-11e5-83f1-aac481e04e25.png)

*Example Rendered markdown page*
![image](https://cloud.githubusercontent.com/assets/6663710/8419171/67730b2e-1e7d-11e5-8723-a28df1a21b62.png)

*Another Example*
![image](https://cloud.githubusercontent.com/assets/6663710/8419192/810011ae-1e7d-11e5-82a2-72a1f1868af5.png)

parlarjb commented on README.md:
```diff
@@ -60,9 +60,9 @@ A breaking change is one that *requires* users of the changed component to modif
 
  - [Roadmap](./guides/roadmap.md)
- - [Contribution Guide](./CONTRIBUTING.md)
+ - [Contribution Guide](./guides/contributing.md)
```
You've got to back this out, based on the comment I left here https://github.com/rackerlabs/encore-ui/pull/1071#discussion_r33589265

parlarjb replied:
And you have to back out the associated `git mv`

TheNando replied:
Because I created a route based controller to display these files, I am going to keep the path as '/guides' but I will add an additional copy task to render these files from the root directory but put them in the guide path along with the other rendered markdown files. Same thing for the one below.

parlarjb commented on README.md:
```diff
@@ -60,9 +60,9 @@ A breaking change is one that *requires* users of the changed component to modif
+ - [Contribution Guide](./guides/contributing.md)
  - [Versioning Guide](./guides/versioning.md)
- - [Changelog](./CHANGELOG.md)
+ - [Changelog](./guides/changelog.md)
```
Same with this, back it out please

parlarjb replied:
And you have to back out the associated `git mv`

parlarjb commented on demo/demo.js:
```diff
@@ -1,3 +1,11 @@
+            name: breadcrumbTitle || ''
+        }]);
+    }
+}
```
You're my hero :clap:

parlarjb commented on demo/overview.html:
```diff
@@ -1,3 +1,72 @@
+                <h4 class="title sm">Getting Started</h4>
+
+                <a href="/#/guides/getting-started">Using EncoreUI In Your Own Project</a>
+                <p>How to begin a project that uses the components and styles provied by EncoreUI</p>
```
`provided`

TheNando replied:
That's the...American spelling?

glynnis commented on demo/overview.html:
```diff
@@ -1,3 +1,72 @@
-    <div class="component-description clear" ng-include="'readme.html'"></div>
+<rx-page title="'EncoreUI Overview'" layout="column" layout-fill>
+
+    <p class="clear">EncoreUI is a library of reusable AngularJS widgets for internal Racksapce projects.</p>
```
Typo: "Racksapce" -> "Rackspace"

glynnis commented on demo/overview.html:
```diff
@@ -1,3 +1,72 @@
+                <a href="/#/guides/">Test Docs</a>
+                <p>API documentation for our Page Objects</p>
+
+                <a href="/#/guides/">Unit Test Coverage</a>
```
Is `/#/guides/` The correct location for all these?

glynnis commented on demo/overview.html:
```diff
@@ -1,3 +1,72 @@
+                <p>List of what features and updates are included in each release</p>
+
+                <a href="/#/guides/">Common Questions (not yet implemented)</a>
+                <p>Answers to common questions about the tech stack EncoreUI utilized</p>
```
`utilizes`

TheNando commented:
@parlarjb I've rebased this gain, for easy merging ;)

parlarjb commented on demo/overview.html:
```diff
@@ -1,3 +1,72 @@
+            <div>
+                <h4 class="title sm">APIs &amp; Metrics</h4>
+
+                <a href="/#/guides/">JS Docs</a>
```
Needs a link

parlarjb commented on demo/overview.html:
```diff
@@ -1,3 +1,72 @@
+                <a href="/#/guides/">JS Docs</a>
+                <p>API documentation for our directives, services, factories, and filters</p>
+
+                <a href="/#/guides/">Test Docs</a>
```
Needs a link

parlarjb commented on demo/overview.html:
```diff
@@ -1,3 +1,72 @@
+                <a href="/#/guides/">Test Docs</a>
+                <p>API documentation for our Page Objects</p>
+
+                <a href="/#/guides/">Unit Test Coverage</a>
```
Needs a link

parlarjb commented:
Relative links aren't working. From the Overview page, go to "Introduction to Testing", and from that page click on "Details on this are explained below".

parlarjb commented:
Go to `http://localhost:9001/#/guides/testing`, and down to the "Convenience Page Objects" section. It has a link to "Component Scaffolding". In the Markdown, this is a relative URL, but in the generated HTML it's a link out to github.com. Likely our markdown conversion stuff needs an update for this

parlarjb commented on demo/assets/generated_demo.css:
```diff
@@ -123,3 +134,26 @@ pre code {
   border-radius: 0;
 }
+<<<<<<< HEAD
+=======
```
Tsk tsk tsk

parlarjb commented on demo/assets/generated_demo.css:
```diff
@@ -123,3 +134,26 @@ pre code {
+  margin: 10px 10px 10px 23px;
+  text-indent: -13px;
+}
+>>>>>>> docs(overview): implement newly organzied overview page
```
Tsk tsk tsk

megalord commented:
From the breadcrumbs in the screenshots, I noticed that the pages are part of the "Overview" section.  However, the nav item is just a link - it isn't expandable.  Is the only way to access the child pages to go to the main overview page first and then click on the link in the page?  I'm wondering if the overview pages should be children of the overview nav item.

megalord commented on grunt-tasks/options/copy.js:
```diff
@@ -36,29 +105,24 @@ module.exports = {
             dest: 'demo/readme.html'
```
Can the `demoreadme` task be merged with the new `demomarkdown` one? It seems like you could just add this object into the `files` array for that task.

TheNando replied:
Done. I'll upload shortly.

megalord commented on grunt-tasks/options/copy.js:
```diff
@@ -1,5 +1,74 @@
+    var id = text.toLowerCase().replace(/[^\w ]/g, '').replace(/ /g, '-');
+    var openTag = '<h%level% id="%id%">'
+        .replace('%id%', id)
+        .replace('%level%', level + (transform[level] || ''));
```
I think [`_.template`](http://devdocs.io/lodash/index#template) would make this a little cleaner.

TheNando replied:
I agree. Plus, I like the ES6 syntax better than what lodash proposes.

TheNando replied:
Done.

megalord commented on demo/guides/guideController.js:
```diff
@@ -0,0 +1,48 @@
+            var anchorPos = $location.$$url.indexOf('#');
+
+            if (anchorPos !== -1) {
+                var anchorId = $location.$$url.substr(anchorPos + 1);
```
It's worthwhile avoiding "private" properties when possible in my opinion.  Would `$location.path()` also not make it easier to get the `anchorId`?

TheNando replied:
Cool. Thanks. I cleaned this all up.

megalord commented on demo/guides/guideController.js:
```diff
@@ -0,0 +1,48 @@
+        vm.activate = activate;
+        vm.guide = $routeParams.guide;
+        vm.scrollTo = scrollTo;
+        vm.setTitle = setTitle;
```
Does `setTitle()` need to be on the `$scope`?  I don't see it being used.

TheNando replied:
Yeah. That was sloppy of me. I removed it.

megalord commented on demo/guides/guideController.js:
```diff
@@ -0,0 +1,48 @@
+            title.remove();
+
+            // If the page is cached, the next time this runs, it will grab the incorrect data
+            $templateCache.remove('guides/rendered/' + vm.guide + '.html');
```
If I understand correctly, this handles the case where a markdown file is edited locally, but the changes don't show up?  Wouldn't a (cache-clearing) refresh solve that problem?

TheNando replied:
Not exactly. For full context, read below. TL;DR version is that without this line, the 2nd time you visit a page without refreshing, it will insert the wrong title.

So the controller loads the file based on the route sent to it. Then, if using `rxPage`, you set the `title` attribute which does a lot of things:
* sets the page title in the tab bar
* injects a page template, which includes:
   * breadcrumbs
   * a hidden notification bar
   * a header with the page title

The alternative was manually setting all that stuff but the header, since it wasn't really needed. But, that seemed very redundant, so instead I opted to hijack the title by binding it to a variable instead a string literal. I somewhat arbitrarily decided to pick the first found header as the title because it seemed appropriate. But, because the router executes the controller before the page has been loaded, I had to bind setting the title to the `onload` function.

Turns out that when you go back to the overview, either using the browser back button or the breadcrumbs, if you visit any page _again_, the `onload` will still execute, but, for some reason `document` will still contain the Overview page's HTML content, not the content of the page you visited for the second time.

So, after lots of guessing, I figured it had something to do with the template cache and accidentally discovered that it didn't happen if you removed the page after you were finished changing the title.

megalord replied:
Ohhh very interesting.  Thanks for the detailed explanation!

megalord commented on demo/overview.html:
```diff
@@ -1,3 +1,81 @@
+            <div>
+                <h4 class="title sm">Getting Started</h4>
+
+                <a href="/#/guides/using-encoreui">Using EncoreUI In Your Own Project</a>
```
Will a leading `/` break these links in the `github.io` domain?

TheNando replied:
I'm not sure. Does anyone else know about this?

parlarjb replied:
I can't recall. Let's remove to be safe.

halkon replied:
It will:

![image](https://cloud.githubusercontent.com/assets/1437154/8625181/ac0755fa-2702-11e5-8de5-f1ae527cb905.png)

TheNando replied:
Done.

parlarjb commented:
From the "EncoreUI Developer Setup" page, scroll down to "Modifying Existing Modules", and hover over the "Add tests!" link in the first bullet. The link there is to the `testing.md` markdown file

parlarjb commented:
In "How to contribute", in the "PR Steps" section, the first four bullets of "Step 3" are being blank

parlarjb commented:
In "Introduction to Testing", in the "Convenience Page Objects" section, the "Component Scaffolding" link in the second paragraph is pointing to the markdown file.

There seems to be a problem with files in `guides/` doing relative references to other files in `guides/`

glynnis commented:
Can we move the icons before the links in the docs section (so they more closely match page actions)?

<img width="311" alt="screen shot 2015-07-13 at 11 35 51 am" src="https://cloud.githubusercontent.com/assets/1529366/8654995/d63abb52-2953-11e5-8fb9-f4f7c548d5f4.png">

<img width="241" alt="screen shot 2015-07-13 at 11 40 38 am" src="https://cloud.githubusercontent.com/assets/1529366/8655011/f04a5be2-2953-11e5-81ed-10591bf71d07.png">

TheNando commented:
@glynnis, @parlarjb I implemented your suggestions and / or fixed the issues you noted. I also captured all the links from every page and classified them based on execution path through the link rewriter, ensuring that I was aware of every case and that they each were being handled. I had to rework a few things, but the main issue I was having was that the links were already being rewritten by the parent markdown call. I removed that and I was able to simplify the renderer and catch every case.

@megalord 
>From the breadcrumbs in the screenshots, I noticed that the pages are part of the "Overview" section. However, the nav item is just a link - it isn't expandable. Is the only way to access the child pages to go to the main overview page first and then click on the link in the page? I'm wondering if the overview pages should be children of the overview nav item.

To my knowledge, there will not be any sub-links on the nav. Though, all the routes in the URL are properly parsed by the controller, so you can bookmark any page and revisit it with no problem, including anchor links.

@glynnis, related to the point that @megalord brought up, should the Overview nav item stay blue with the arrow even when visiting sub-links? Right now, it is not.

CITguy commented on demo/guides/guideViewer.html:
```diff
@@ -0,0 +1,3 @@
+<rx-page title=vm.title>
```
Should the `title` value be without quotes?

TheNando replied:
You are correct. The surprising thing is that it worked without it. I'll put them in though.

glynnis commented:
@TheNando Current state (only highlighted when you're actually on the overview page itself, not its children) is correct.

Droogans commented:
@TheNando did you update master before you rebased? I'm seeing `./utils/visual-regression/clone.sh: line 6: travis_retry: command not found`, which was [fixed already](https://github.com/rackerlabs/encore-ui/pull/1093).

I really do need you to get a valid visual regression run against this branch because there's an open issue in node-canvas that breaks installs on Mac OS (so I'm currently stuck) and I'd like to confirm whether it'll install on Linux.

CITguy commented:
Where are we with this PR?

TheNando commented:
I have rebased this again and run unit tests and midways. Everything is working locally. If it passes build will someone please merge this?

parlarjb commented:
I will Nando, I will. I'll do this for you.

TheNando commented:
:+1:

TheNando commented:
![sad](https://wnursports.files.wordpress.com/2013/05/tumblr_li7un4ifyi1qeb0lyo1_400.jpg)

CITguy commented on demo/assets/demo.less:
```diff
@@ -174,31 +184,53 @@ table.component-attributes {
 
```
please add `.markdown` here as an additional namespace for code styling

CITguy replied:
This is needed to apply the code styling (see #1057 and #1070) for the generated markdown pages.

CITguy commented on demo/assets/demo.less:
```diff
@@ -174,31 +184,53 @@ table.component-attributes {
 
 .component-description,
 .component-demo {
```
Please add `.markdown` here as an additional namespace for code styling.  This is needed to apply the code styling (see #1057 and #1070) for the generated markdown pages.

CITguy replied:
### without `.markdown` (wall of text)
![screen shot 2015-08-10 at 11 28 40 am](https://cloud.githubusercontent.com/assets/545605/9176801/096f525c-3f53-11e5-8fe5-606eb2951996.png)

### with `.markdown`
![screen shot 2015-08-10 at 11 29 04 am](https://cloud.githubusercontent.com/assets/545605/9176808/0db1573e-3f53-11e5-809f-b4b36fd946de.png)

TheNando replied:
Good catch. I also thought it looked a little bare.

glynnis commented:
@TheNando I noticed that the top-most nav section is different in this PR than in some of the InVision screens for this story (in InVision there's no "EncoreUI" section, and the links now in the "APIs & Metrics" sections are no longer in the nav). Is that something that could be tackled with this PR, or will it be done separately?

TheNando commented:
@glynnis Originally, this branch was just to get the markdown viewer working, but it had since creeped into much more functionality. Might as well add the rest.

TheNando commented:
@glynnis On second though, I don't think it will be a trivial change. Let's get this branch merged first, then we can move on to the rest of the reorg.

glynnis commented:
@TheNando Cool, sounds good to me :+1: Also LGTM. When is it time to merge?

![giphy](https://cloud.githubusercontent.com/assets/1529366/9206750/48f471b4-402f-11e5-9f66-c51400ab5f1a.gif)

TheNando commented:
![merge](https://i.imgflip.com/peh3d.jpg)

CITguy commented:
LGTM
